src/Microsoft.Data/PrimitiveColumnContainer.cs

```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
                 if (Buffers.Count == 0)
                 {
                     Buffers.Add(new DataFrameBuffer<T>());
+                    NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
                 }
                 DataFrameBuffer<T> lastBuffer = Buffers[Buffers.Count - 1];
                 if (lastBuffer.Length == lastBuffer.MaxCapacity)
                 {
                     lastBuffer = new DataFrameBuffer<T>();
                     Buffers.Add(lastBuffer);
+                    NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
                 }
                 int allocatable = (int)Math.Min(length, lastBuffer.MaxCapacity);
                 lastBuffer.EnsureCapacity(allocatable);
+                DataFrameBuffer<byte> lastNullBitMapBuffer = NullBitMapBuffers[NullBitMapBuffers.Count - 1];
+                lastNullBitMapBuffer.EnsureCapacity(allocatable);
                 lastBuffer.Length = allocatable;
+                lastNullBitMapBuffer.Length = allocatable;
                 length -= allocatable;
                 Length += lastBuffer.Length;
             }
         }
 
-        public void Append(T value)
+        public void Append(T? value)
         {
             if (Buffers.Count == 0)
             {
                 Buffers.Add(new DataFrameBuffer<T>());
+                NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
             }
             DataFrameBuffer<T> lastBuffer = Buffers[Buffers.Count - 1];
             if (lastBuffer.Length == lastBuffer.MaxCapacity)
             {
                 lastBuffer = new DataFrameBuffer<T>();
+                Buffers.Add(lastBuffer);
+                NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
             }
-            lastBuffer.Append(value);
+            lastBuffer.Append(value ?? default);
+            SetBit(Length, value.HasValue ? true : false);
             Length++;
         }
 
+        public bool IsValid(long index) => NullCount == 0 || GetBit(index);
+
+        private void SetBit(long index, bool value)
+        {
+            if (index < 0 || index > Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+            // First find the right bitMapBuffer
+            int bitMapIndex = (int)(index / Buffers[0].MaxCapacity);
+            Debug.Assert(NullBitMapBuffers.Count > bitMapIndex);
+            DataFrameBuffer<byte> bitMapBuffer = NullBitMapBuffers[bitMapIndex];
+
+            // Set the bit
+            index -= bitMapIndex * Buffers[0].MaxCapacity;
+            int bitMapBufferIndex = (int)index / 8;
```

Comment:
```suggestion
            int bitMapBufferIndex = (int)index >> 3;    // index / 8
```
The bit-shift will produce [better code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAbgFgAoYgZkoCZyBhcgb0fIGVWASwB2GcgFkAFABsIogObkxAExgIAlOQC8APnLSxGTWo3kA9OQAc9BoPL9BLFePIA5OQuVmtug0bipqLqCOR6Bsx2AL5AA=)

JIT will substitute division by a power of two only for unsigned ints / longs. So instead of the bit-shift it [could also be](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAbgFgAoYgZkoCZyBhcgb0fIGVWASwB2GcgFkAFABsIogObkxAExgIAlOQC8APnLSxGTWo3kA9OQAc9BoPL9BLFePIA5OQuVmtug0bipqLqCOR6Bsx2Dk4CLsZSVF5KrqHa+obGmtLSAK5ZvpY2mnYAvkA===)
```suggestion
            int bitMapBufferIndex = (int)((uint)index / 8);
```

You can take whatever you prefer :wink:
I would take the latter.